REVIEW ONLY: Improve Embedded Planet's initial port of MCUboot to Mbed OS

LDong-Arm opened the pull request:
Based on https://github.com/JuulLabs-OSS/mcuboot/pull/791

Current supported:
  * One image only, with primary slot in internal flash and secondary slot being an mbed::BlockDevice instance (any internal or
    external storage) provided by the user
  * Mbed TLS crypto backend - TinyCrypt is not currently supported until we have implemented its NRG interface

See mcuboot/boot/mbed/mbed_lib.json for supported and unsupported/untested features, mcuboot/boot/mbed/README.md for usage.

AGlass0fMilk committed:
Started porting to Mbed library configuration system

AGlass0fMilk committed:
Updated mbed_lib.json to enable FlashIAP block device functionality

AGlass0fMilk committed:
Adding initial infrastructure for mbed-based mcuboot bootloader

AGlass0fMilk committed:
Filled out flash algorithm and headers needed by mcuboot

AGlass0fMilk committed:
Working out flash API...

AGlass0fMilk committed:
Updated flash block opening device handling

AGlass0fMilk committed:
Fixed invalid read sizes causing an assert

AGlass0fMilk committed:
Updated so encryption keys are application-defined

AGlass0fMilk committed:
Fixed extern declaration of enc_priv_key

AGlass0fMilk committed:
Created configuration to allow for library-only builds of mcuboot. This lets an application call mcuboot functions without a conflicting main declaration.

AGlass0fMilk committed:
Ignore cypress from Mbed-OS builds

AGlass0fMilk committed:
Updated to build with latest mcuboot

AGlass0fMilk committed:
Reverted logging configuration (mcuboot only checks to see if it's defined...)

AGlass0fMilk committed:
Removed specified-name macros from all config options

AGlass0fMilk committed:
Removed target overrides. These should be specified by user building the bootloader

AGlass0fMilk committed:
Added application header size configuration parameter

AGlass0fMilk committed:
Enable primary slot validation by default

AGlass0fMilk committed:
Ignore flash map backend during non-bootloader build

AGlass0fMilk committed:
Refactored secondary_bd declaration to a function

AGlass0fMilk committed:
Update default image header size and init flash areas

AGlass0fMilk committed:
Update flash backend to build for bootloader or application properly

AGlass0fMilk committed:
Stopgap fix for multi image slot API

AGlass0fMilk committed:
Fix address inconsistencies between bootloader/application builds

AGlass0fMilk committed:
Clarifying address assignment and making it possible to build bootloader to avoid collision with primary app header

LDong-Arm commented:
Looks like `.github/workflows/` (from the upstream MCUboot) requires "Signed-off-by" in the commit message.

evedon commented on boot/mbed/app_enc_keys.c:
```diff
@@ -0,0 +1,83 @@
+#include <bootutil/sign_key.h>
+
+// TODO - Remove Zephyr references
+// TODO - Make framework option of multiple public keys
```
These two TODOs should be removed.
Zephyr references should be removed too, which is essentially the next block comment.

evedon commented on boot/mbed/include/flash_map_backend/secondary_bd.h:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
```
Are these copyrights correct?

LDong-Arm replied:
Good catch on this one, I'll remove those lines since secondary_bd is mbed-specific.
For the other files that are based on existing configs/code for other platforms in the mcuboot repo, I kept the copyrights in the original files

evedon commented on boot/mbed/include/mcuboot_config/mcuboot_config.h:
```diff
@@ -0,0 +1,118 @@
+#define MCUBOOT_SIGN_ED25519
+#endif
+
+// TODO - Support for Cryptocell310?
```
Remove the TODO

evedon commented on boot/mbed/include/mcuboot_config/mcuboot_config.h:
```diff
@@ -0,0 +1,118 @@
+#if (MCUBOOT_CRYPTO_BACKEND == MBEDTLS)
+#define MCUBOOT_USE_MBED_TLS
+#elif (MCUBOOT_CRYPTO_BACKEND == TINYCRYPT)
+#define MCUBOOT_USE_TINYCRYPT
```
Since tinycrypt is not currently supported, should we issue a compilation error?

LDong-Arm replied:
I'll add an error for this one, but most other configs that are unsupported/untested do not have a check currently. Will add a check for this one in particular for now.

evedon commented on boot/mbed/include/mcuboot_config/mcuboot_config.h:
```diff
@@ -0,0 +1,118 @@
+ */
+
+#ifndef MCUBOOT_IMAGE_NUMBER
+#define MCUBOOT_IMAGE_NUMBER 1
```
We should also check that MCUBOOT_IMAGE_NUMBER is not greater than 1 since we don't have support for multiple images.

LDong-Arm replied:
As we don't support multiple images, this is not configurable in mbed_lib.json. I'll remove the `#ifndef`.

evedon commented on boot/mbed/include/mcuboot_config/mcuboot_config.h:
```diff
@@ -0,0 +1,118 @@
+ * The function priority is:
+ *
+ *    MCUBOOT_LOG_ERR > MCUBOOT_LOG_WRN > MCUBOOT_LOG_INF > MCUBOOT_LOG_DBG
+ */
```
I would remove this block comment

evedon commented on boot/mbed/mbed_lib.json:
```diff
@@ -0,0 +1,159 @@
+            "value": null
+        },
+        "hardware-key": {
+            "help": "Use hardware key (NOT TESTED)",
```
That's a nice way to document it

evedon commented on boot/mbed/mbed_lib.json:
```diff
@@ -0,0 +1,159 @@
+            "value": null
+        }
+    }
+}
```
New line missing

evedon commented on boot/mbed/mcuboot_main.cpp:
```diff
@@ -0,0 +1,62 @@
+
+    uint32_t address = rsp.br_image_off + rsp.br_hdr->ih_hdr_size;
+
+    tr_info("Booting firmware image at 0x%x", address);
```
There is no longer a need to add `\n`?

LDong-Arm replied:
No need for stdlib-based printing - mbed-trace uses `puts()` from stdlib (like `printf()`) instead of raw serial. Some flushing mechanism is in place I guess.

LDong-Arm replied:
Even though it is printed, it doesn't add a newline... need to add the workaround back.

LDong-Arm replied:
Added back

evedon commented on boot/mbed/src/flash_map_backend.cpp:
```diff
@@ -0,0 +1,231 @@
+                MCUBOOT_LOG_ERR("Read failed: %d", ret);
+                return ret;
+            }
+            memcpy(dst, buffer, remainder);
```
Should be `memcpy(dst + len, buffer, remainder);`

LDong-Arm replied:
Good catch!

evedon commented on boot/mbed/mbed_lib.json:
```diff
@@ -0,0 +1,159 @@
+        "read-granularity": {
+            "help": "Granularity of read operations, in bytes. It should be no less than the read size of all storage devices (no need to be multiples of them).",
+            "macro_name": "MCUBOOT_READ_GRANULARITY",
+            "value": null
```
You mentioned that MCUboot requires a granularity of 16 bytes.
I think you need this when using a secondary block device from QSPIF. Can you explain how you use this configuration?

LDong-Arm replied:
Do you mean explain it in "help" here?

evedon replied:
I meant in a reply comment in the PR. Or perhaps I should review the applications related PR.

LDong-Arm replied:
It's used in `flash_area_read()` - it creates a buffer `uint8_t buffer[MCUBOOT_READ_GRANULARITY]`, reads data into the buffer, then memcpy's whatever is needed to the input buffer.

evedon reviewed (changes requested):
The README.md file is excellent.
I gave a few comments left related to further cleanup. 

I think there is an issue in `flash_area_read` when MCUBOOT_READ_GRANULARITY is set.

LDong-Arm commented on boot/mbed/include/mcuboot_config/mcuboot_config.h:
```diff
@@ -0,0 +1,118 @@
+ */
+#define MCUBOOT_USE_FLASH_AREA_GET_SECTORS
+
+/* Default maximum number of flash sectors per image slot; change
+ * as desirable. */
+#if !defined(MCUBOOT_MAX_IMG_SECTORS)
+#define MCUBOOT_MAX_IMG_SECTORS 128
+#endif
```
I'll remove this since it's a "required" config in mbed_lib.json - it's always defined

LDong-Arm reviewed (commented):
@evedon Thanks for the review, I'll fix the comments and upload demo apps that use this PR asap.

LDong-Arm commented on boot/mbed/src/flash_map_backend.cpp:
```diff
@@ -0,0 +1,231 @@
+
+    // Note: The address must be aligned to bd->get_read_size(). If MCUBOOT_READ_GRANULARITY
+    // is defined, the length does not need to be aligned.
+#ifdef MCUBOOT_READ_GRANULARITY
```
@evedon This is the config usage you asked about

evedon commented:
Great work @LDong-Arm

LDong-Arm reviewed (commented):
Thanks @evedon 🙂

LDong-Arm commented:
Updated bootloader app: https://github.com/LDong-Arm/mbed-mcuboot-demo/pull/1
Update flow demo: https://github.com/LDong-Arm/mbed-mcuboot-blinky/pull/1

TTornblom committed:
boot: Fix FIH_LABEL/FIH_CALL macros for IAR

The IAR toolchain does not handle the "%=" template. Emulate it with
a combination of __LINE__ and __COUNTER__

Signed-off-by: TTornblom <thomas.tornblom@iar.com>

LDong-Arm commented:
Rebased; now swap update is enabled by default.

sigvartmh committed:
bootutil: ecdsa: Fix CC310 for nRF devices

This contains fixes for CC310 for nRF devices using the nrfxlib.
Removed some const for compilation warnings and added `NUM_ECC_BYTES`
define.

Signed-off-by: Sigvart Hovland <sigvart.m@gmail.com>

LDong-Arm committed:
Improvements from Mbed OS team

Some of key changes:
  * Addresses and sizes of each flash region are directly set
    in the configuration instead of deducted from Mbed OS macros.
    The whole storage layout of a bootloader should be exactly
    known, predictable and within the developer's control.
  * Reworked configurations
  * Integrated mbed-trace
  * Added a brief boot/mbed/README.md
  * General formatting improvements and clean-up of unused files
    and comments.

Signed-off-by: Evelyne Donnaes <evelyne.donnaes@arm.com>
Signed-off-by: Lingkai Dong <lingkai.dong@arm.com>
